04_basic_struct/lkm_example.c

```diff
@@ -0,0 +1,146 @@
+// SPDX-License-Identifier: GPL-2.0 OR BSD-3-Clause
+
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/printk.h>
+#include <linux/moduleparam.h>
+#include <linux/kernel.h>
+#include <linux/sysfs.h>
+#include <linux/types.h>
+#include <linux/list.h>
+#include <linux/slab.h>
+
+MODULE_AUTHOR("Ivan Kryvosheia");
+MODULE_DESCRIPTION("Test module for Linux Kernel ProCamp");
+MODULE_LICENSE("Dual BSD/GPL");
+MODULE_VERSION("0.01");
+
+struct my_container {
+	struct list_head list;
+	int some_member;
+};
+
+struct my_obj {
+	struct list_head node;
+	char *dyn_data;
+};
+
+
+static ssize_t procamp_show(struct kobject *kobj,
+		struct kobj_attribute *attr, char *buf);
+static ssize_t procamp_store(struct kobject *kobj,
+		struct kobj_attribute *attr, const char *buf, size_t count);
+
+struct kobject *kobj_inst;
+
+static struct kobj_attribute procamp_attr = {
+	.attr = { .name = "list", .mode = 0666 },
+	.show	= procamp_show,
+	.store	= procamp_store
+};
+struct my_container container;
+
+int init_sysfs_attr(void)
+{
+	int rc = 0;
+
+	//create a directory into the /sys/kernel directory
+	kobj_inst = kobject_create_and_add("gl_procamp", kernel_kobj);
+
+	if (!kobj_inst)
+		return -1;
+
+	rc = sysfs_create_file(kobj_inst, &procamp_attr.attr);
+
+	if (rc) {
+		pr_debug("failed to create the foo file in /sys/kernel/kobject_example\n");
+		return -1;
+	}
+
+	return rc;
+}
+
+void deinit_sysfs_attr(void)
+{
+	if (kobj_inst)
+		kobject_put(kobj_inst);
```

Comment:
need to call sysfs_remove_file before kobject_put

Reply:
thanks for that